### PR TITLE
Changelog GitHub Action

### PR DESCRIPTION
**Pull request by DEATHB4DEFEAT**
## GitHub Environment Variables needed
Sub-headers have a linked guide on where to put these variables, items have an example in (parenthesis)

### [Secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository)
- BOT_TOKEN - A Personal Access Token with `Contents - Write` access

### [Vars](https://docs.github.com/en/actions/learn-github-actions/variables#defining-configuration-variables-for-multiple-workflows)
- CHANGELOG_BRANCH - Which branch to push the changelog updates to (master)
- CHANGELOG_DIR - The path to the changelog file (Resources/changelog/changelog.yml)
- CHANGELOG_MESSAGE - The commit comment (Automatic Changelog Update)
- CHANGELOG_USER - For git metadata, can probably be whatever
- CHANGELOG_EMAIL - For git metadata, use the bot account's email

### CODE REVIEW

File: Tools/changelog/package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "axios": "^1.3.4",
+    "js-yaml": "^4.1.0"
+  }
+}
```
Comment (Just-a-Unity-Dev) [inline Tools/changelog/package.json:16]: no npm hell please

File: Tools/changelog/changelog.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+// Using a token is optional, but will greatly reduce the chance of hitting the rate limit
+if (process.env.GITHUB_TOKEN) axios.defaults.headers.common["Authorization"] = `Bearer ${process.env.GITHUB_TOKEN}`;
+
+(async () => {
+    // Debug
+    console.log(`PR Number: ${process.env.PR_NUMBER}`);
+    console.log(`Using Token: ${process.env.GITHUB_TOKEN != undefined && process.env.GITHUB_TOKEN != ""}`);
+    console.log(`Repository: ${process.env.GITHUB_REPOSITORY}`);
+    console.log(`Changelog Directory: ${process.env.CHANGELOG_DIR}`);
+    console.log("\n");
```
Comment (Just-a-Unity-Dev) [inline Tools/changelog/changelog.js:14]: if it's debug then why isn't it being removed in prod mode

File: .github/workflows/changelog.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      - name: Setup Node
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18.x
+
+      - name: Install Dependencies
+        run: |
+          cd "${{ github.action_path }}"
+          cd Tools
+          cd changelog
```
Comment (Just-a-Unity-Dev) [inline .github/workflows/changelog.yml:39]: can't you just do `cd "${{ github.action_path }}"/Tools/changelog`???

File: .github/workflows/changelog.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          npm install
+        shell: bash
+
+      - name: Generate Changelog
+        run: |
+          cd "${{ github.action_path }}"
+          cd Tools
+          cd changelog
```
Comment (Just-a-Unity-Dev) [inline .github/workflows/changelog.yml:47]: above

**Review by Just-a-Unity-Dev — CHANGES_REQUESTED** [review]
take a look at how SS14.Changelog works internally please

**Comment by Just-a-Unity-Dev** [comment]
1. how does this handle clearing the CL?
2. are you not able to use the `github-actions` bot to commit things? i've written a gh action once and i *have* used said bot
3. you do not 3+ lines to `cd`, or to `console.log` to the terminal

File: Tools/changelog/changelog.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        // time = "2023-03-28T00:00:00.0000000+00:00";
+    }
+    console.log(`Merge Time: ${time}`);
+    console.log("\n");
+
+    // Construct changelog entry
+    const entry = {
+        author: author,
+        changes: entries,
+        id: parseInt(process.env.PR_NUMBER),
```
[the remark this answers is not in the document]
Comment (DEATHB4DEFEAT) [reply Tools/changelog/changelog.js:100]: [looks like the SS14.Changelog uses pr number for id](https://github.com/space-wizards/SS14.Changelog/blob/master/SS14.Changelog/Controllers/WebhookController.cs#L184)

File: Tools/changelog/package.json (hunk 1 above)
Comment (DEATHB4DEFEAT) [reply Tools/changelog/package.json:16]: what

Comment (Just-a-Unity-Dev) [reply Tools/changelog/package.json:16]: > what

why are you using npm for a c# project, cant you just use something else like python that has basic stuff (`requests`) built in

Comment (DEATHB4DEFEAT) [reply Tools/changelog/package.json:16]: I don't know python and I've used JavaScript a lot previously

**Comment by Just-a-Unity-Dev** [comment]
when there's around 100(?) cl entries, the script removes the oldest one so that the cl isn't 5 gb

File: Tools/changelog/changelog.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Just-a-Unity-Dev) [reply Tools/changelog/changelog.js:87]: > [looks like the SS14.Changelog uses pr number for id](https://github.com/space-wizards/SS14.Changelog/blob/master/SS14.Changelog/Controllers/WebhookController.cs#L184)

image

if so then explain the ID on this image

**Comment by DEATHB4DEFEAT** [comment]
*you can do that manually*, but I'll see what I can do

**Comment by Just-a-Unity-Dev** [comment]
> _you can do that manually_, but I'll see what I can do

i can also enforce people to make the CLs manually and close this PR

**Comment by DEATHB4DEFEAT** [comment]
> i can also enforce people to make the CLs manually and close this PR

That is awful, I've tried doing it

File: Tools/changelog/changelog.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (DEATHB4DEFEAT) [reply Tools/changelog/changelog.js:87]: hmm

what's with wrong using the PR number?

[the remark this answers is not in the document]
Comment (moonheart08) [reply Tools/changelog/changelog.js:87]: The real changelog bot supports incorporating changelog data from more than just PRs, and runs on any push to a branch that includes a "changelog part"
The IDs are important as they tell the game when you last read the changelog and must be continuous.

**Comment by DEATHB4DEFEAT** [comment]
- Uses the next largest ID it can for the CL
- Uses GitHub username if no author is provided
- Ignores comments

**Comment by DEATHB4DEFEAT** [comment]
> please don't use js for this simple script

why

**Comment by Just-a-Unity-Dev** [comment]
I'm currently working on a python version of this script, I refuse to merge this due to us already using python for most scripts. It doesn't make sense for us to bundle a node project in our C# repo.

**Pull request closed without merging**